File: frontend/src/api/mutations/smtp.ts
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { api } from '@/api/client';
import type { SmtpStatus } from '../hooks/useSmtp';

/** Тело POST /settings/smtp/config — пустой password означает «не менять». */
export interface SmtpConfigRequest {
  host: string;
  port: number;
  encryption: 'tls' | 'ssl' | 'none';
  username: string;
  password: string;
  from_email: string;
  from_name: string;
}

/** Результат POST /settings/smtp/test. */
export interface SmtpTestResult {
  sent_to: string;
  last_test_at: string;
}

/** POST /settings/smtp/config — сохранение параметров SMTP-сервера. */
export function useSmtpSaveConfig() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (data: SmtpConfigRequest): Promise<SmtpStatus> => {
      const res = await api.post('/settings/smtp/config', data);
      return res.data as SmtpStatus;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['settings', 'smtp'] });
    },
  });
}

/** POST /settings/smtp/test — отправка тестового письма на указанный адрес. */
export function useSmtpTest() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (data: { to: string }): Promise<SmtpTestResult> => {
      const res = await api.post('/settings/smtp/test', data);
      return res.data as SmtpTestResult;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['settings', 'smtp'] });
    },
  });
}

/** POST /settings/smtp/disconnect — сброс SMTP-настроек. */
export function useSmtpDisconnect() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (): Promise<{ message: string }> => {
      const res = await api.post('/settings/smtp/disconnect');
      return res.data as { message: string };
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['settings', 'smtp'] });
    },
  });
}

File: backend/app/schemas/smtp.py
"""Схемы SMTP-настроек (request-модели).

Статус (GET) отдаётся сервисом как обычный dict (пароль в нём отсутствует по
построению — отдельная response-модель не нужна, чтобы случайно не «протащить»
секрет). Валидация значений (порт/режим/почта) — в сервисе (единый источник
ошибок для теста и сохранения).
"""

from pydantic import BaseModel


class SmtpConfigRequest(BaseModel):
    host: str
    port: int
    encryption: str = "ssl"
    username: str = ""
    # Пароль write-only: пусто при уже сохранённой конфигурации → старый пароль
    # сохраняется; наружу (в GET-статусе) не возвращается никогда.
    password: str = ""
    from_email: str
    from_name: str = ""


class SmtpTestRequest(BaseModel):
    to: str
